### PR TITLE
[spark] adjust paimon spark structure

### PR DESCRIPTION
**Pull request by YannByron**
### Purpose

To modify the spark structure as below (It is similar to Apache Gluten https://github.com/apache/incubator-gluten):
image

1. adjust the `spark-common` module to the bottom layer, and make the `spark3-common` and `spark4-common` modules depend on it;
2. extract the separate `spark-ut` module;
3. the specific spark modules (like `spark-3.5` or `spark-4.0`) can directly depend on the `sparkx-common` module with the corresponding major version (eg: `spark3.5` -> `spark3-common`, `spark4.0` -> `spark4-common`);
4. the specific spark-test modules will depend on `spark-ut` and `sparkx-common` modules;
5. after this adjustment, we use the `SparkShim` interface and the SPI mechanism to solve the incompatibility problem, instead of overwriting the classes with the same class name;
6. fix some fragile UTs to make them robust in all spark versions.

Linked issue: close #xxx

### CODE REVIEW

**Comment by YannByron** [comment]
@Zouxxyy @JingsongLi please take a look.

File: paimon-spark/paimon-spark3-common/src/main/scala/org/apache/spark/sql/paimon/shims/Spark3Shim.scala [hunk 1]
```diff
@@ -16,57 +16,53 @@
...
-    val delegate: SparkParserInterface
+  override def createSparkParser(delegate: ParserInterface): ParserInterface = {
+    new PaimonSparkSqlExtensionsParser(delegate)
   }
 
-  abstract class ArrayData extends SparkArrayData {}
-
-  abstract class InternalRow extends SparkInternalRow {}
+  override def createSparkInternalRow(rowType: RowType): SparkInternalRow = {
+    new Spark3InternalRow(rowType)
```
Comment (Zouxxyy) [inline paimon-spark/paimon-spark3-common/src/main/scala/org/apache/spark/sql/paimon/shims/Spark3Shim.scala:42]: Maybe we can use a unified style, such as unifying `PaimonSpark3SqlExtensionsParser` `Spark3InternalRow`,
 or `PaimonSparkSqlExtensionsParser` `SparkInternalRow`.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/spark/sql/paimon/shims/SparkShimLoader.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+package org.apache.spark.sql.paimon.shims
+
+import java.util.ServiceLoader
+
+import scala.collection.JavaConverters._
+
+/** Load a [[SparkShim]]'s implementation. */
+object SparkShimLoader {
+
+  private var sparkShim: SparkShim = _
```
[the remark this answers is not in the document]
Comment (YannByron) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/spark/sql/paimon/shims/SparkShimLoader.scala:28]: OK

File: paimon-spark/paimon-spark3-common/src/main/scala/org/apache/spark/sql/paimon/shims/Spark3Shim.scala (hunk 1 above)
Comment (YannByron) [reply paimon-spark/paimon-spark3-common/src/main/scala/org/apache/spark/sql/paimon/shims/Spark3Shim.scala:42]: Right，I choose the first style.

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/data/SparkInternalRow.scala [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+package org.apache.paimon.spark.data
+
+import org.apache.paimon.types.RowType
+
+import org.apache.spark.sql.catalyst.InternalRow
+import org.apache.spark.sql.paimon.shims.SparkShimLoader
+
+abstract class SparkInternalRow extends InternalRow {
```
[the remark this answers is not in the document]
Comment (YannByron) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/paimon/spark/data/SparkInternalRow.scala:26]: For this kind of interface, I prefer to keep the definition and implementation separate.

**Review by Zouxxyy — APPROVED** [review]
+1 and left some comments

**Review by JingsongLi — APPROVED** [review]
+1

File: paimon-spark/paimon-spark-common/src/main/scala/org/apache/spark/sql/paimon/shims/SparkShimLoader.scala [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    if (sparkShim == null) {
+      sparkShim = loadSparkShim()
+    }
+    sparkShim
+  }
+
+  private def loadSparkShim(): SparkShim = {
+    val shims = ServiceLoader.load(classOf[SparkShim]).asScala
+    if (shims.size != 1) {
+      throw new IllegalStateException("No available spark shim here.")
```
[the remark this answers is not in the document]
Comment (YannByron) [reply paimon-spark/paimon-spark-common/src/main/scala/org/apache/spark/sql/paimon/shims/SparkShimLoader.scala:37]: OK

**Pull request merged**